spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go

```diff
@@ -0,0 +1,123 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+// [START spanner_capture_gfe_metric]
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	spanner "cloud.google.com/go/spanner/apiv1"
+	gax "github.com/googleapis/gax-go/v2"
+	sppb "google.golang.org/genproto/googleapis/spanner/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+// OpenCensus Tag, Measure and View.
+var (
+	KeyMethod    = tag.MustNewKey("grpc_client_method")
+	GFELatencyMs = stats.Int64("cloud.google.com/go/spanner/gfe_latency",
+		"Latency between Google's network receives an RPC and reads back the first byte of the response", "ms")
+	GFELatencyView = view.View{
+		Name:        "cloud.google.com/go/spanner/gfe_latency",
+		Measure:     GFELatencyMs,
+		Description: "Latency between Google's network receives an RPC and reads back the first byte of the response",
+		Aggregation: view.Distribution(0.0, 0.01, 0.05, 0.1, 0.3, 0.6, 0.8, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0, 10.0, 13.0,
+			16.0, 20.0, 25.0, 30.0, 40.0, 50.0, 65.0, 80.0, 100.0, 130.0, 160.0, 200.0, 250.0,
+			300.0, 400.0, 500.0, 650.0, 800.0, 1000.0, 2000.0, 5000.0, 10000.0, 20000.0, 50000.0,
+			100000.0),
+		TagKeys: []tag.Key{KeyMethod}}
+)
+
+func queryWithGFELatency(w io.Writer, db string, projectID string) error {
+	ctx := context.Background()
+	client, err := spanner.NewClient(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// Register OpenCensus views.
+	err = view.Register(&GFELatencyView)
+	if err != nil {
+		return err
+	}
+
+	// Create OpenCensus Stackdriver exporter.
+	sd, err := stackdriver.NewExporter(stackdriver.Options{
+		ProjectID: projectID,
+	})
+	if err != nil {
+		return err
+	}
+	// It is imperative to invoke flush before your main function exits
+	defer sd.Flush()
+
+	// Start the metrics exporter
+	sd.StartMetricsExporter()
+	defer sd.StopMetricsExporter()
+
+	// Create a session.
+	req := &sppb.CreateSessionRequest{Database: db}
+	session, err := client.CreateSession(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	// Execute a SQL query and retrieve the GFE server-timing header in gRPC metadata.
+	req2 := &sppb.ExecuteSqlRequest{
+		Session:   session.Name,
+		Sql:       `SELECT SingerId, AlbumId, AlbumTitle FROM Albums`,
+		QueryMode: sppb.ExecuteSqlRequest_PROFILE,
```

Comment:
I thought PROFILE must be set in query_stats sample and not here. I mean this is what I did in Java samples.

Reply:
You're right it isn't needed in this sample. Removed.